src/compiler/binder.ts

```diff
@@ -352,6 +374,272 @@ namespace ts {
             blockScopeContainer = savedBlockScopeContainer;
         }
 
+        function shouldSaveReachabilityState(n: Node): boolean {
+            return n.kind === SyntaxKind.SourceFile || n.kind === SyntaxKind.ModuleBlock || isFunctionLike(n);
+        }
+
+        function bindWithReachabilityChecks(node: Node): void {
+            let savedReachabilityState: Reachability;
+            let savedLabelStack: Reachability[];
+            let savedLabels: Map<number>;
+            let savedImplicitLabels: number[];
+            let savedHasExplicitReturn: boolean;
+
+            let saveState = shouldSaveReachabilityState(node);
+            if (saveState) {
+                savedReachabilityState = currentReachabilityState;
+                savedLabelStack = labelStack;
+                savedLabels = labelIndexMap;
+                savedImplicitLabels = implicitLabels;
+                savedHasExplicitReturn = hasExplicitReturn;
+
+                currentReachabilityState = Reachability.Reachable;
+                hasExplicitReturn = false;
+                labelStack = labelIndexMap = implicitLabels = undefined;
+            }
+
+            if (!bindReachableStatement(node)) {
+                forEachChild(node, bind);

```

Comment:
I would move the `forEachChild` into the `default` clause of the switch statement in `bindReachableStatement` and just have that function return void.